Added a dummy code of conduct

dominic-mulligan-arm opened the pull request:
Note that the Github web interface only allows one to select between two different Codes of Conducts for some reason, neither of which are the ones we agreed with the CCC during project adoption, but creating a Code of Conduct file through the web interface is the only way to avoid it thinking that there is no such file!  As a result, I'm adding this file which will be updated with the real Code of Conduct once merged.

hugovincent commented on Dockerfile:
```diff
@@ -11,29 +11,113 @@
+## fetch and build the SDK
+    wget https://download.01.org/intel-sgx/sgx-linux/2.9.1/distro/ubuntu18.04-server/sgx_linux_x64_sdk_2.9.101.2.bin \
+    && chmod a+x ./sgx_linux_x64_sdk_2.9.101.2.bin \
+    && echo "yes" | ./sgx_linux_x64_sdk_2.9.101.2.bin \
```
Not your fault but do they really distribute the SDK as a self-extracting executable? Yuck!

basmaelgaabouri replied:
in their [documentation](https://download.01.org/intel-sgx/sgx-linux/2.9.1/docs/Intel_SGX_Installation_Guide_Linux_2.9.1_Open_Source.pdf) that's still the official recommendation, but we can change that to building the executable from their git repo, should I change that?

hugovincent replied:
I wasn't requesting a change, just making a comment – if that's what they recommend it's fine (just a little surprising to see in a security-oriented project).

ShaleXIONG replied:
They also publish source code [here](https://github.com/intel/linux-sgx) @hugovincent.

dominic-mulligan-arm commented:
Great work, @basmaelgaabouri.  I've just merged @geky's pull request ahead of this one, as it was apparently sat in the queue for two weeks without anybody noticing.  From looking at his changes, it shouldn't be too hard to rebase your changes on top of his: he just adds `absname` to the Makefile in various spots around `VERACRUZ_ROOT` so that relative paths work.  This looks good to me, though I'll wait for a review from @ShaleXIONG before merging.

ebeasant-arm commented on README.md:
```diff
@@ -0,0 +1,31 @@
+# Veracruz Docker image
+
+## Requierments
```
Very nitty nit, typo: ## Requirements

ebeasant-arm commented on Makefile:
```diff
@@ -22,26 +22,27 @@ development: build
 ci: Dockerfile.ci
-	docker build --build-arg USER=root --build-arg UID=0 -t $(VERACRUZ_DOCKER_IMAGE) -f $<  .
+	DOCKER_BUILDKIT=1 docker build --squash --build-arg USER=root --build-arg UID=0 --build-arg TEE=sgx -t $(VERACRUZ_DOCKER_IMAGE)_sgx -f $<  .
+	DOCKER_BUILDKIT=1 docker build --squash --build-arg USER=root --build-arg UID=0 --build-arg TEE=tz -t $(VERACRUZ_DOCKER_IMAGE)_tz -f $<  .
```
Where a multi-layer build is used (in this case for trustzone), I might have expected to see a --cache-from=build_tz_base and perhaps a --cache-from=build_tee here?

basmaelgaabouri replied:
to my understanding, it's done implicitly since some stages depend on the one before, I could be wrong though ..

ebeasant-arm replied:
If it shows up as a problem after the commit, we can revisit it, I guess :-)

basmaelgaabouri replied:
sounds good to me

dominic-mulligan-arm commented on Dockerfile:
```diff
@@ -11,29 +11,113 @@
+    wget https://download.01.org/intel-sgx/linux-2.5/ubuntu18.04-server/libsgx-enclave-common-dbgsym_2.5.101.50123-bionic1_amd64.ddeb \
+    && dpkg -i libsgx-enclave-common-dbgsym_2.5.101.50123-bionic1_amd64.ddeb \
+    && rm libsgx-enclave-common-dbgsym_2.5.101.50123-bionic1_amd64.ddeb; \
+# more cleaning needed
```
Is this comment stray?  Or is it a TODO?

basmaelgaabouri replied:
oops! .. forgot it there

dominic-mulligan-arm commented on README.md:
```diff
@@ -0,0 +1,31 @@
+## Requirements
+
+We use `squash` docker experimental feature to help reduce docker image size, to enable this feature:
+- run: 
```
Are these instructions Linux-specific?  We should probably note that, if they are.

basmaelgaabouri replied:
yes they are, I've updated readme

basmaelgaabouri commented on Dockerfile:
```diff
@@ -115,97 +198,84 @@ RUN apt-get update && apt-get install -y \
-    && git submodule update --init libm \
-    && cd ../rust \
-    && git submodule update --init src/stdarch src/llvm-project 
+    && rm -rf rust
```
fixed

basmaelgaabouri commented on Dockerfile:
```diff
@@ -11,29 +11,113 @@
+RUN apt-get update && apt-get install --no-install-recommends -y \
+    libcurl4-openssl-dev libprotobuf-dev python \
+    libssl-dev build-essential \
+    wget make curl unzip ca-certificates \
```
fixed

basmaelgaabouri commented on Dockerfile:
```diff
@@ -115,97 +198,84 @@ RUN apt-get update && apt-get install -y \
-RUN if [ "$USER" != "root" ] ; then . /home/$USER/.cargo/env ; else . /root/.cargo/env; fi \
+    if [ "$USER" != "root" ] ; then . /home/$USER/.cargo/env ; else . /root/.cargo/env; fi \
     && rustup component add rust-src \
     && cargo install xargo --force \
```
fixed

dominic-mulligan-arm commented:
@ShaleXIONG to confirm, the changes in the main Veracruz repository to support this change can be merged both before or after this is approved, right?  That is, there is no dependency on either one being merged first?

If so, @dreemkiller, when you are happy with these changes please go ahead and merge these changes.

ShaleXIONG commented:
> @ShaleXIONG to confirm, the changes in the main Veracruz repository to support this change can be merged both before or after this is approved, right? That is, there is no dependency on either one being merged first?
> 
> If so, @dreemkiller, when you are happy with these changes please go ahead and merge these changes.

@dominic-mulligan-arm . There is no direct dependency. I will wait this go through first, and then I will replace the image in AWS. Therefore we can let CI check if [the veracruz pull request](https://github.com/veracruz-project/veracruz/pull/52) is good. However, without the new change, the old Veracruz repo will NOT be able to compile in the new docker image. That is, the new image has a 'smaller' environment.

dreemkiller commented:
When I build for SGX, I get `execvp: clang-check: Permission denied` when building psa-attestation. Is anyone else seeing this?

ShaleXIONG commented:
> When I build for SGX, I get `execvp: clang-check: Permission denied` when building psa-attestation. Is anyone else seeing this?

Short answer: I did not see this before.
I used `make sgx` in my run. Did you manually compile psa-attestaion and get that problem? Can you provide a bit more information, so @basmaelgaabouri and I can check what goes wrong in the image.

dreemkiller commented:
I used `make sgx-sinaloa-test`

dreemkiller commented:
Ok, I think I have more germane questions.
When I enter the docker container with my username (`docker exec -u <USERNAME> -it veracruz bash`), and I run `make sgx-sinaloa-test`, I get the following error:

> make -C data-generators/string-generator/
make[2]: Entering directory '/work/veracruz/sdk/data-generators/string-generator'
cargo run -- --input_file hello-world-1.txt
info: syncing channel updates for 'nightly-2020-05-07-x86_64-unknown-linux-gnu'
info: latest update on 2020-05-07, rust version 1.45.0-nightly (1836e3b42 2020-05-06)
info: downloading component 'cargo'
info: downloading component 'rust-std'
info: downloading component 'rustc'
info: installing component 'cargo'
info: Defaulting to 500.0 MiB unpack ram
info: installing component 'rust-std'
info: installing component 'rustc'
    Updating crates.io index
warning: spurious network error (2 tries remaining): failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_9tRRf3': Permission denied; class=Os (2)
warning: spurious network error (1 tries remaining): failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_UtETaF': Permission denied; class=Os (2)
error: failed to get `clap` as a dependency of package `string-generator v0.3.0 (/work/veracruz/sdk/data-generators/string-generator)`
Caused by:
  failed to load source for dependency `clap`
Caused by:
  Unable to update registry `https://github.com/rust-lang/crates.io-index`
Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`
Caused by:
  failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_MtgSpi': Permission denied; class=Os (2)
Makefile:29: recipe for target 'hello-world-1.dat' failed

This appears to be a problem with which user is active when rust was installed in the Dockerfile. Previously, the` $USER `was active. In the new version, the active user is `root`. Thus, previously cargo and rust were installed in /home/$USER/.cargo, but are now installed in `/usr/local/cargo`, which is not writable by the `$USER` user.
I'm not sure how to resolve this without possibly breaking the CI system, so I need some advice on why this change was made and how we might resolve it.

ShaleXIONG commented:
> Ok, I think I have more germane questions.
> When I enter the docker container with my username (`docker exec -u <USERNAME> -it veracruz bash`), and I run `make sgx-sinaloa-test`, I get the following error:
> 
> > make -C data-generators/string-generator/
> > make[2]: Entering directory '/work/veracruz/sdk/data-generators/string-generator'
> > cargo run -- --input_file hello-world-1.txt
> > info: syncing channel updates for 'nightly-2020-05-07-x86_64-unknown-linux-gnu'
> > info: latest update on 2020-05-07, rust version 1.45.0-nightly (1836e3b42 2020-05-06)
> > info: downloading component 'cargo'
> > info: downloading component 'rust-std'
> > info: downloading component 'rustc'
> > info: installing component 'cargo'
> > info: Defaulting to 500.0 MiB unpack ram
> > info: installing component 'rust-std'
> > info: installing component 'rustc'
> > Updating crates.io index
> > warning: spurious network error (2 tries remaining): failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_9tRRf3': Permission denied; class=Os (2)
> > warning: spurious network error (1 tries remaining): failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_UtETaF': Permission denied; class=Os (2)
> > error: failed to get `clap` as a dependency of package `string-generator v0.3.0 (/work/veracruz/sdk/data-generators/string-generator)`
> > Caused by:
> > failed to load source for dependency `clap`
> > Caused by:
> > Unable to update registry `https://github.com/rust-lang/crates.io-index`
> > Caused by:
> > failed to fetch `https://github.com/rust-lang/crates.io-index`
> > Caused by:
> > failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_MtgSpi': Permission denied; class=Os (2)
> > Makefile:29: recipe for target 'hello-world-1.dat' failed
> 
> This appears to be a problem with which user is active when rust was installed in the Dockerfile. Previously, the`$USER`was active. In the new version, the active user is `root`. Thus, previously cargo and rust were installed in /home/$USER/.cargo, but are now installed in `/usr/local/cargo`, which is not writable by the `$USER` user.
> I'm not sure how to resolve this without possibly breaking the CI system, so I need some advice on why this change was made and how we might resolve it.

One solution is to grant permission to $USER for accessing /usr/local/cargo, when building up the image. @basmaelgaabouri Can you make the change?

dominic-mulligan-arm commented:
@basmaelgaabouri, @ShaleXIONG, @dreemkiller, @ebeasant-arm: this now looks ready for merge.  Any final comments/blockers on this?

dominic-mulligan-arm committed:
Added a dummy code of conduct

Note that the Github web interface only allows one to select between two different Codes of Conducts for some reason, neither of which are the ones we agreed with the CCC during project adoption, but creating a Code of Conduct file through the web interface is the only way to avoid it thinking that there is no such file!  As a result, I'm adding this file which will be updated with the real Code of Conduct once merged.
